bindings/pyroot/pythonizations/python/ROOT/_pythonization/_roofit/_roodatahist.py

```diff
@@ -50,3 +50,190 @@ def plotOn(self, *args, **kwargs):
         # The keywords must correspond to the CmdArg of the `plotOn` function.
         args, kwargs = _kwargs_to_roocmdargs(*args, **kwargs)
         return self._plotOn(*args, **kwargs)
+
+    @property
+    def shape(self):
+        import ROOT
+        import numpy as np
+
+        shape = []
+
+        var_is_category = self._var_is_category()
+
+        for i_var, var in enumerate(self.get()):
+            shape.append(var.numBins())
+
+        def get_len_from_shape():
+            n_size_check = 1
+            for n in shape:
+                n_size_check = n_size_check * n
+            return n_size_check
+
+        assert get_len_from_shape() == len(self)
+
+        return tuple(shape)
+
+    def __len__(self):
+        return self.numEntries()
+
+    def _to_array(self, buffer):
+        import numpy as np
+
+        # check if buffer is nullptr (for some reason comparing with ROOT.nullptr doesn't work)
+        try:
+            buffer[0]
+        except ReferenceError:
+            return None
+        a = np.copy(np.frombuffer(buffer, dtype=np.float64, count=len(self)))
+        return a.reshape(self.shape)
+
+    def _var_is_category(self):
+        """ """
+        import ROOT
+
+        out = []
+
+        for i_var, var in enumerate(self.get()):
+            if isinstance(var, ROOT.RooAbsReal):
+                out.append(False)
+            elif isinstance(var, ROOT.RooAbsCategory):
+                out.append(True)
+            else:
+                raise TypeError("Vaiables in RooDataHist should either be RooAbsReal or RooAbsCategory.")
+
+        return out
+
+    def _weight_error_low(self):
+        """Returns the low weight errors as numpy arrays."""
+
+        return self._to_array(self.wgtErrLoArray())
+
+    def _weight_error_high(self):
+        """Returns the high weight errors as numpy arrays."""
+        return self._to_array(self.wgtErrHiArray())
+
+    def _weight_error(self):
+        """Returns the low and high weight errors as numpy arrays."""
+        return self.weight_error_low(), self.weight_error_high()
+
+    def _weights_squared_sum(self):
+        """Returns the sum of squared weights that were used to fill each bin."""
+        return self._to_array(self.sumW2Array())
+
+    @staticmethod
+    def from_numpy(hist_weights, variables, bins, ranges=None, weights_squared_sum=None, name=None, title=None):
```

Comment:
Should we name it `from_numpy` and `to_numpy` or `FromNumpy` and `ToNumpy` to comply with the C++ naming (and also other non-C++ cases such as `RDataFrame.AsNumpy`)?

Reply:
Good point, but let me argue why I'd rather stick with `from_numpy` and `to_numpy`.

  * It gives a warm fuzzy feeling to the Python user that the function names follow the established Python convention (maybe sounds ridiculous but I think it is psychologically very important because it makes ROOT feel more native to Python)
  * It highlights the distinction between pythonized C++ functions and pure Python functions
  * We have already advertised these names in the PPP and ACAT:
      * https://indico.cern.ch/event/1061658/contributions/4461926/attachments/2286164/3887473/210722_PPP_RooFit_Pythonization.pdf
      * https://indico.cern.ch/event/855454/contributions/4596763/attachments/2356539/4021590/Accelerating%20RooFit%20with%20GPUs.pdf
  * For RooDataSet, we already used the `snake_case` names: https://github.com/root-project/root/pull/9346
  * At this point, the `to_numpy` and `from_numpy` names are almost conventional in the Python ecosystem, see e.g. Pandas or PyTorch. So a user would try `d.to_numpy()` intuitively, just like they would try `d.shape` and assume that it works

By the way, RooFit follows `camelCase`, unlike the rest of ROOT. So if we try to comply with the C++ names, it would rather be `toNumpy` and `fromNumpy`.